sdk/core/dma/dma_compartment.cc

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#define MALLOC_QUOTA 0x100000
+
+#include <cheri.hh>
+#include <compartment-macros.h>
+#include <utils.hh>
+#include <thread_pool.h>
+
+#if __has_include(<platform-dma.hh>)
+#	include <platform-dma.hh>
+#endif
+
+// Import some useful things from the CHERI namespace.
+using namespace CHERI;
+
+using thread_pool::async;
+/**
+ *  creating a global heap below.
+ *  todo: we assume that this heap capability can serve 
+ *  for different compartments, irrespective of 
+ *  the origin because this is a driver code?
+*/
+DECLARE_AND_DEFINE_ALLOCATOR_CAPABILITY(dmaDriverHeap, 8144);
+
+SObjStruct *dmaHeap;
+
+Ibex::PlatformDMA platformDma;
+
+void free_dma(uint32_t *sourceAddress, uint32_t *targetAddress)
+{
+    int sourceStatus = heap_free(dmaHeap, sourceAddress);
+    int targetStatus = heap_free(dmaHeap, targetAddress);
+
+    /**
+     *  ideally, free should not fail if claim was successful
+     *  todo: but in case, assert a Debug::Assert() for heap_free() later
+     */    
+}
+
+int dma_compartment(uint32_t *sourceAddress, uint32_t *targetAddress, uint32_t lengthInBytes,
+                        uint32_t sourceStrides, uint32_t targetStrides, uint32_t byteSwapAmount)
+{    
+    /**
+     *  return if sufficient permissions are not present 
+     *  and if not long enough 
+     */
+
+    if (!check_pointer<PermissionSet{Permission::Load, Permission::Global}>(sourceAddress, lengthInBytes) ||
```

Comment:
Check pointer should happen after claim. The tag can be cleared by a concurrent free if they happen the other way around.